js/src/chat/chat-external-link.ts

```diff
@@ -0,0 +1,177 @@
+import { LitElement, html } from "lit"
+import { property } from "lit/decorators.js"
+
+import { LightElement } from "../utils/_utils"
+
+declare global {
+  interface Window {
+    shinychat_always_open_external_links?: boolean
+  }
+}
+
+export class ChatExternalLinkDialog extends LightElement {
+  @property() url = ""
+
+  private static instance: ChatExternalLinkDialog | null = null
+  private dialog: HTMLDialogElement | null = null
+  private resolvePromise: ((value: boolean) => void) | null = null
+
+  static getInstance(): ChatExternalLinkDialog {
+    if (!ChatExternalLinkDialog.instance) {
+      ChatExternalLinkDialog.instance = document.createElement(
+        "shinychat-external-link-dialog",
+      ) as ChatExternalLinkDialog
+      document.body.appendChild(ChatExternalLinkDialog.instance)
+    }
+    return ChatExternalLinkDialog.instance
+  }
+
+  connectedCallback(): void {
+    super.connectedCallback()
+    this.createDialog()
+  }
+
+  disconnectedCallback(): void {
+    super.disconnectedCallback()
+    this.dialog?.remove()
+    this.dialog = null
+  }
+
+  render() {
+    return html``
+  }
+
+  /**
+   * Shows a modal dialog asking for confirmation to open an external link
+   * @param url The URL to open
+   * @returns A promise that resolves to true if the user confirmed, false otherwise
+   */
+  async showConfirmation(url: string): Promise<boolean> {
+    // If the user has opted to always open external links, resolve immediately
+    if (window.shinychat_always_open_external_links) {
+      return Promise.resolve(true)
+    }
+
+    this.url = url
+
+    // Update the URL in the dialog
+    const linkUrlElement = this.dialog?.querySelector(".link-url")
+    if (linkUrlElement) {
+      linkUrlElement.textContent = url
+    }
+
+    return new Promise<boolean>((resolve) => {
+      this.resolvePromise = resolve
+
+      try {
+        this.dialog?.showModal()
+      } catch (err) {
+        // If showModal fails, resolve with true to fallback to default behavior
+        resolve(true)
+      }
+    })
+  }
+
+  private createDialog(): void {
```

Comment:
I'd prefer this broken down into 2 methods: `createDialog(): HTMLElement` and `attachDialogListeners(x: HTMLElement): void`. Then do all state mutation at the top-level in `connectedCallback()`

Reply:
I considered this but the event listeners aren't really _state_. Creating the `<dialog>` element and adding its event listeners are part of the same process and you wouldn't want to do them separately.